File: Week_02/java/src/_107_BinaryTreeLevelOrderTraversalII.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class _107_BinaryTreeLevelOrderTraversalII {
    static class Bfs {
        public List<List<Integer>> levelOrderBottom(TreeNode root) {
            if (root == null) return Collections.emptyList();
            LinkedList<List<Integer>> ret = new LinkedList<>();
            LinkedList<TreeNode> queue = new LinkedList<>();
            queue.offer(root);
            while (!queue.isEmpty()) {
                int size = queue.size();
                List<Integer> list = new ArrayList<>(size);
                while (size-- > 0) {
                    TreeNode node = queue.poll();
                    list.add(node.val);
                    if (node.left != null) queue.offer(node.left);
                    if (node.right != null) queue.offer(node.right);
                }
                ret.offerFirst(list);
            }
            return ret;
        }
    }

    static class Dfs {
        public List<List<Integer>> levelOrderBottom(TreeNode root) {
            if (root == null) return Collections.emptyList();
            ArrayList<List<Integer>> ret = new ArrayList<>();
            dfs(root, 0, ret);
            Collections.reverse(ret);
            return ret;
        }

        void dfs(TreeNode root, int level, List<List<Integer>> ret) {
            if (root == null) return;
            if (level == ret.size()) ret.add(new ArrayList<>());
            ret.get(level).add(root.val);
            dfs(root.left, level + 1, ret);
            dfs(root.right, level + 1, ret);
        }
    }

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }
}

File: Week_02/kotlin/src/_107_BinaryTreeLevelOrderTraversalII.kt
import java.util.*

class _107_Dfs {
    fun levelOrderBottom(root: TreeNode?): List<List<Int>> {
        root ?: return emptyList()
        val ret = mutableListOf<MutableList<Int>>()
        fun dfs(node: TreeNode, level: Int) {
            if (level == ret.size) ret += mutableListOf<Int>()
            ret[level].add(node.`val`)
            node.left?.let { dfs(it, level + 1) }
            node.right?.let { dfs(it, level + 1) }
        }
        dfs(root, 0)
        return ret.apply { reverse() }
    }
}

class _107_Bfs {
    fun levelOrderBottom(root: TreeNode?): List<List<Int>> {
        root ?: return emptyList()
        val queue = LinkedList<TreeNode>().apply { offer(root) }
        val ret = LinkedList<List<Int>>()
        while (queue.isNotEmpty()) {
            var size = queue.size
            val list = mutableListOf<Int>()
            while (size-- > 0) {
                val node = queue.poll()
                list += node.`val`
                node.left?.let { queue.offer(it) }
                node.right?.let { queue.offer(it) }
            }
            ret.offerFirst(list)
        }
        return ret
    }
}